New CI infrastructure to support private ELK per test

AmitShaul opened the pull request:
## Motivation
added new CI infra
there's another PR open #2197, the difference is that in #2197 there is a code duplication of the tests (tests_elk) folder which doesn't exists here.

## Changes
Node pool for each test
ELK cluster per each test
Both node pool and ELK cluster are deployed using new pytest fixtures

## Overview
**new test:**
When starting a new test a namespace will be created for that test and that test only.
First an ELK cluster will be created over that namespace, ELK is built from Fluent-bit, Logstash, Kibana and Elasticsearch.
After setting the ELK cluster we wait for ES and Logstash to be ready, this process might take a couple of minutes.
When ELK is ready we add the node pool that will serve the Spacemesh nodes, this node pool is necessary in order to get all of the resources needed (for our SM clients) in advance so our deployment will happen in a shorter and more predictable time.
The process of creating\deleting\altering a node pool is blocking since only one action can be made simultaneously which means that if someone just started a test and wants to create a node pool he will have to wait for previous started action to end.
Node pool creation\deletion is polling every 10 seconds to see if it can start and in case it can't (for the reason mentioned above) it  creates a single "Error" message for the first time it tries, this message can be ignored for it will start once GCP is ready to take another action.

**Teardown**:
After a test was done a process of teardown will start in the opposite order, first the tests' node pool will be deleted (where most of the tests' resources are) and after that the namespace will be deleted deleting the ELK with it.

**Logging**:
After a test was finished logs will be dumped to the main ES server located on 'logging' namespace.
Logs will be dumped under two conditions:
1. The test has failed.
2. 'is_dump: True' was added to the tests' config file.

In any other case those logs will be lost with the namespace deletion.
Until namespace is deleted logs are accessible in the tests' private ELK, the IPs can be listed by running:
'kubectl get services -n YOUR_NAMESPACE'

AmitShaul committed:
WIP, added elk yaml files

sudachen committed:
elastic search k8s fix

sudachen committed:
storageclass fix

AmitShaul committed:
WIP, added elk conf files, added fixture to run elk, need to remove former elastic handler

AmitShaul committed:
Merge branch 'develop' of github.com:spacemeshos/go-spacemesh into elk

AmitShaul committed:
WIP ES context to work with new elk, fixed a bug in kibana deployment text file

AmitShaul committed:
WIP added logstass, added debugging code that should be removed

sudachen committed:
Update elastic_stateful.yaml

sudachen committed:
Update service.yaml

Do not publish elastic to the public Internet

sudachen committed:
Merge pull request #2154 from spacemeshos/sudachen-patch-1

Update elastic_stateful.yaml

sudachen committed:
some fixes for ELK

AmitShaul committed:
WIP elk works with kubernetes metadata, still sleeping in mining test

AmitShaul committed:
elk integrated and working, should add dumping of elastic data to a central elastic on teardown

AmitShaul committed:
Merge branch 'develop' of github.com:spacemeshos/go-spacemesh into elk

AmitShaul committed:
fixed a bug when printing unsuccessful miners in query_message

sudachen committed:
logstash and elastic must use SSD storage not HDD

AmitShaul committed:
new CI is working, changed logstash stateful

AmitShaul committed:
Merge branch 'develop' of github.com:spacemeshos/go-spacemesh into elk

antonlerner committed:
trying a fix

AmitShaul committed:
WIP all tests but hares tests are passing, commit for testing with Antons fix

AmitShaul committed:
Merge branch 'sync_atxs' of github.com:spacemeshos/go-spacemesh into elk_hare

AmitShaul committed:
WIP using elasticdump to dump ES data, not all issues resolved with es_dump

AmitShaul committed:
added support for ES dumping using ES API

AmitShaul committed:
deprication with extensions/v1beta1, changed deployment to work with app/v1 instead, changed k8s automation config files to support the newer version

AmitShaul committed:
changed ES main URL

AmitShaul committed:
fixed oracle k8s object with older and deprecated extensions/v1beta1

AmitShaul committed:
Merge branch 'develop' of github.com:spacemeshos/go-spacemesh into elk_hare

AmitShaul committed:
fixed namespace bug in filebeat, NAMESPACE was switched with a static namespace. handled exceptions in es_reindex(). fixed NAMESPACE replace regex with negative look behind for _ char.

AmitShaul committed:
added timeout for es_dump + removed v1beta api when deleting deployment

AmitShaul committed:
merged support for fluent bit, added sleep time before starting p2p tests

AmitShaul committed:
Merge branch 'develop' of github.com:spacemeshos/go-spacemesh into elk_hare

AmitShaul committed:
fixed logstash input bug, fixed es_dump bug

AmitShaul committed:
[*] namespace support for fluent bit, now fluent bit will only collect logs from its own namespace.
[*] kube deployment files with modified namespaces deletion happens right after file was loaded to memory.
[*] added tolerance to elastic_logging k8s files.
[*] added pod affinity to fluent bit daemonset
[*] added namespace labels to k8s client, bootstrapoet and poet files (only for those with no ss)

AmitShaul committed:
merged dev_elk
[*] wait for elk to be ready
[*] changed k8s resources request and limit to be equal
[*] fixed main ES toleration
[*] fixed Fluent bit to collect only sm files as input
[*] replaced podAffinity with nodeSelector in fluent bit
[*] added support for namespacing when applying statefulset k8s file

AmitShaul committed:
added support for node pool per test
need to fix delayed nodes

AmitShaul commented:
bors try

AmitShaul committed:
fixed makefile and ci.yml to work with the new flow

AmitShaul commented:
bors try

AmitShaul committed:
ci.yml fixed ci-stage2 needs to be the new tests

AmitShaul commented:
bors try

AmitShaul commented:
bors try-

AmitShaul committed:
ci.yml fixed ci-stage2 status

AmitShaul commented:
bors try

AmitShaul committed:
Makefiles dockerbuild-test-elk, fixed bug with dockerbuild-test-elk - changed DockerFileTestsElk to DockerFileTests

AmitShaul commented:
bors try

AmitShaul committed:
fixed genesis timing
[*] lowered genesis time for add\remove nodes, late nodes and hare tests
[*] fixed number of nodes when deploying node pool in NodePoolDep

AmitShaul commented:
bors try

AmitShaul committed:
fixed bugs
[*] test mining raised genesis delta to 50
[*] test_p2p raised pods to 60

AmitShaul commented:
bors try

AmitShaul committed:
[*] added support for ES reinstallation incase of an exception
[*] added handler for k8s files applyment
[*] adjust genesis delta in add\remove nodes and test_bs

AmitShaul commented:
bors try

AmitShaul committed:
[*] changed add_curl fixture not to depand on setup_bootstrap
[*] changed add_curl to happen before deploying clients and bootstrap
[*] raised create_pod timeout to 70
[*] raised remove_nodes genesis_delta to 70

AmitShaul commented:
bors try

AmitShaul committed:
DEBUGGING added sleeping time after ES and logstash timeout exception for debugging purposes

AmitShaul commented:
bors try

AmitShaul committed:
removed debugging from wait_for_minimal_elk_cluster_ready, raised es timeout to 240 and ls timeout to 180

AmitShaul commented:
bors try

web3rover committed:
run on custom runner

web3rover committed:
changed runner name

web3rover commented:
bors try

web3rover commented:
bors try

AmitShaul committed:
removed es removal incase of an exception, raised logstash timeout to 240, after fixing github actions timeout should be no more than 120

AmitShaul committed:
Merge branch 'elk_hare' of github.com:spacemeshos/go-spacemesh into elk_hare

web3rover committed:
use ubuntu-latest runner

web3rover committed:
Merge branch 'elk_hare' of github.com:spacemeshos/go-spacemesh into elk_hare

AmitShaul commented:
bors try

AmitShaul commented:
bors try-

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul committed:
fixed 500 exception by adding retries when posting a curl request to a client pod

AmitShaul commented:
bors try

AmitShaul commented:
bors try-

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul commented:
bors try-

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul committed:
fixed utils api_call function

AmitShaul committed:
[*] set a new ES service
[*] changed the order of elk and pool fixtures in order to allow node pool to be deleted before ES dumping is happening

AmitShaul commented:
bors try

AmitShaul committed:
raised hare test genesis time from 40 to 55

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul committed:
[*] new and stronger ES, changed the ES main IP under config.py
[*] added a new argument when running pytest tests, is-dump, when set to true logs will be dumped to the main ES server when tests are finished.
[*] removed old commands from Makefile and added is-dump argument when running new pytest elk tests

AmitShaul commented:
bors try

AmitShaul committed:
[*] changed is-dump to be a boolean
[*] added keep-alive header to ES dump request

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul committed:
changed is-dump to be boolean

AmitShaul commented:
bors try

AmitShaul committed:
removed keep-alive header from ES dump

AmitShaul committed:
removed is-dump option, only failed tests will dump logs into main ES server when tests are done

y0sher commented on tests/config.py:
```diff
@@ -3,6 +3,16 @@
+LOGSTASH_CONF_DIR = './elk/logstash/'
+
+ES_USER_LOCAL = "elastic"
+ES_PASS_LOCAL = "gavradon"
```
it's better not to include the password here as plain text

AmitShaul replied:
is base64 adequate?

lrettig replied:
no, we should not store passwords in text files. It should be a GitHub secret attached to the repository and passed in through the environment.

y0sher commented on tests/config.py:
```diff
@@ -3,6 +3,16 @@
+
+ES_USER_LOCAL = "elastic"
+ES_PASS_LOCAL = "gavradon"
+MAIN_ES_URL = "35.197.82.152:9200"
```
also maybe create a k8s service for that like we have now ?

y0sher commented on tests/conftest.py:
```diff
@@ -243,12 +247,36 @@ def save_log_on_exit(request):
+    deployer = NodePoolDep(testconfig)
+    _, time_elapsed = deployer.add_node_pool()
+    yield time_elapsed
+    deployer.remove_node_pool()
```
maybe add some indicator about time passed so we have some feeling of how much time we're waiting ?

y0sher replied:
also wouldn't this code remove the node_pool right after its created?

AmitShaul replied:
added a print for time_elapsed.
and about your question, no.
everything after the yield command is working as a destructor and will run only when this fixture will perform teardown.

AmitShaul committed:
[*] polling for ES documents status after dumping instead of waiting for response from _reindex
[*] fixed execution function to print to screen only with conditions to prevent console flooding.

AmitShaul committed:
added is_dump option, when set to true (t, yes, 1..) logs will be dumped to  main ES even if tests have passed

lrettig commented on .github/workflows/ci.yml:
```diff
@@ -157,9 +159,9 @@ jobs:
       - name: checkout
         uses: actions/checkout@v2
-      - name: hare+mining system tests
-        run: make dockertest-hare-mining
```
Was this test removed? Or replaced?

AmitShaul replied:
we don't run tests together in pairs anymore

lrettig commented on .github/workflows/ci.yml:
```diff
@@ -168,9 +170,9 @@ jobs:
       - name: checkout
         uses: actions/checkout@v2
-      - name: sync+blocks remove node test
-        run: make dockertest-sync-blocks-remove-node
```
This one too, was it removed or replaced? I don't see it anywhere else

AmitShaul replied:
you can see them run separately

lrettig commented on Makefile:
```diff
@@ -227,104 +236,103 @@ DELIM=::
 
 
-dockerrun-p2p:
+dockerrun-p2p-elk:
```
How important is it to rename this and all the other make recipes to add `-elk`? I think you could just modify the recipes to use `dockerbuild-test-elk` without renaming them all no?

AmitShaul replied:
tests before were named without the elk and because I had to change them and keep the old ones I added the "-elk"
I don't think it's something I need to change, if anyone has a strong objection I can consider it.

lrettig commented on tests/analyse.py:
```diff
@@ -53,10 +53,11 @@ def analyze_mining(deployment, last_layer, layers_per_epoch, layer_avg_size, tot
+    assert up_pods == total_pods, f"up pods: {up_pods}, total: {total_pods}"
     # not all nodes created blocks
-    assert total_pods == len(blockmap)
+    assert total_pods == len(blockmap), f"total pods={total_pods}, number of nodes created blocks{len(blockmap)}"
```
```suggestion
    assert total_pods == len(blockmap), f"total pods={total_pods}, number of nodes that created blocks={len(blockmap)}"
```

AmitShaul replied:
fixed

lrettig commented on tests/context.py:
```diff
@@ -21,16 +20,19 @@ def wrapper(*args, **kwargs):
+        self.namespace = namespace
+        self.es_ip = self.get_elastic_ip()
+        self.es = Elasticsearch(
+            self.es_ip, port=9200, timeout=30, max_retries=3, retry_on_timeout=True, http_auth=('elastic', 'gavradon')
```
don't hardcode username and password here either

lrettig commented on tests/convenience.py:
```diff
@@ -21,12 +21,16 @@ def convert_ts_to_datetime(ts):
-        print(f" {tts} seconds left     ", end="\r")
+        print(f" {tts} seconds left             ", end="\r")
         time.sleep(1)
+    print("\033[A \033[A")
```
What does this print? Could you add a comment to make it clear from the code?

AmitShaul replied:
it removed the last printed line.
added a comment

lrettig commented on tests/k8s_handler.py:
```diff
@@ -0,0 +1,197 @@
+            elif dep['kind'] == 'Deployment':
+                k8s_client = client.AppsV1Api()
+                k8s_client.delete_namespaced_deployment(name=name, namespace=namespace)
+            # elif dep['kind'] == 'Service':
```
Why add commented-out code here?

AmitShaul replied:
uncommented

lrettig commented on tests/node_pool_deployer.py:
```diff
@@ -0,0 +1,120 @@
+        self.cluster_name = "spacemesh-cluster-elk"
+        self.zone = "us-west1-a"
+        self.default_config = {
+            "project_name": "spacemesh-198810",
```
This (and the `zone`, above) should probably also not be hardcoded, this should probably be passed in from the env so it's easier to change, or stored in a config file or something

lrettig commented on tests/node_pool_deployer.py:
```diff
@@ -0,0 +1,120 @@
+        total_cpu = total_pods * cpu
+        total_mem = total_pods * mem
+        if total_cpu % 2:
+            print("adding additional CPU, total CPU was not divide by 2, total CPUs:", total_cpu)
```
```suggestion
            print("adding additional CPU, total CPU count not divisible by 2, total CPUs:", total_cpu)
```

AmitShaul replied:
fixed

lrettig commented on tests/node_pool_deployer.py:
```diff
@@ -0,0 +1,120 @@
+            print("adding additional CPU, total CPU was not divide by 2, total CPUs:", total_cpu)
+            total_cpu += 1
+        if total_mem % 512:
+            raise Exception("total memory is not a 512 multiplication")
```
```suggestion
            raise Exception("total memory is not divisible by 512")
```

AmitShaul replied:
fixed

AmitShaul committed:
PR #2234 fixes, added some timing prints and other very minor fixes

AmitShaul committed:
[*] removed ES credentials from code, we now relay that the user will have both ES_USER and ES_PASS environment variables set before running automation code.
[*] duplicate_file_and_replace_phrases to handle a list of phrases to change and a list of values to replace with.

AmitShaul committed:
added ES_USER and ES_PASS to makefile and github actions file
